### PR TITLE
Experimental: Trying to add decorators for conditional regeneration of Data Help appreciated

### PR DESCRIPTION
**Pull request by MrDiver**
This should add decorators to affect the `MobjectStatus` which is used in the renderer in order to not recompute everything in every frame.

### CODE REVIEW

File: manim/mobject/opengl/opengl_mobject.py [hunk 1]
```diff
@@ -103,6 +103,59 @@ class MobjectStatus:
...
+
+
+def affects_points(func) -> Callable[..., _O]:
+    """Decorator for methods that change the points of a :class:`~.Mobject`."""
+
+    def wrapper(self: OpenGLMobject, *args) -> _O:
+        self.status.points_changed = True
+        return func(self, *args)
+
+    return wrapper
```
Comment (jeertmans) [inline manim/mobject/opengl/opengl_mobject.py:156]: Maybe use `functools.wraps` to keep function signature and documentation?

```suggestion
def affects_color(func) -> Callable[..., _O]:
    """Decorator for methods that change the color of a :class:`~.Mobject`."""

    @wraps(func)
    def wrapper(self: OpenGLMobject, *args) -> _O:
        self.status.color_changed = True
        return func(self, *args)

    return wrapper


def affects_position(func) -> Callable[..., _O]:
    """Decorator for methods that change the position of a :class:`~.Mobject`."""

    @wraps(func)
    def wrapper(self: OpenGLMobject, *args) -> _O:
        self.status.position_changed = True
        return func(self, *args)

    return wrapper


def affects_rotation(func) -> Callable[..., _O]:
    """Decorator for methods that change the rotation of a :class:`~.Mobject`."""

    @wraps(func)
    def wrapper(self: OpenGLMobject, *args) -> _O:
        self.status.rotation_changed = True
        return func(self, *args)

    return wrapper


def affects_scale(func) -> Callable[..., _O]:
    """Decorator for methods that change the scale of a :class:`~.Mobject`."""

    @wraps(func)
    def wrapper(self: OpenGLMobject, *args) -> _O:
        self.status.scale_changed = True
        return func(self, *args)

    return wrapper


def affects_points(func) -> Callable[..., _O]:
    """Decorator for methods that change the points of a :class:`~.Mobject`."""

    @wraps(func)
    def wrapper(self: OpenGLMobject, *args) -> _O:
        self.status.points_changed = True
        return func(self, *args)

    return wrapper
```

Comment (MrDiver) [reply manim/mobject/opengl/opengl_mobject.py:156]: Ah, sure! Yeah, I didn't think about documentation by now, but that seems like a good idea!

File: manim/mobject/opengl/opengl_mobject.py [hunk 2]
```diff
@@ -103,6 +103,64 @@ class MobjectStatus:
     points_changed: bool = False
 
 
+_O = TypeVar("_O")
```
Comment (VascoSch92) [inline manim/mobject/opengl/opengl_mobject.py:106]: why _0 ? what is that? which kind of object? I will use a more descriptive name

File: manim/mobject/opengl/opengl_mobject.py [hunk 3]
```diff
@@ -103,6 +103,64 @@ class MobjectStatus:
...
+
+
+def affects_color(func) -> Callable[..., _O]:
```
Comment (JasonGrace2282) [inline manim/mobject/opengl/opengl_mobject.py:109]:
```suggestion
_F = TypeVar("_F", bound=Callable[..., Any])
def affects_color(func: _F) -> _F:
```
See <https://mypy.readthedocs.io/en/stable/generics.html#declaring-decorators>

Comment (jeertmans) [reply manim/mobject/opengl/opengl_mobject.py:109]: @JasonGrace2282 As mentioned a few lines below, I tend to favor the new `ParamSpec` type and usage, e.g.,:

```python
from typing import Callable, TypeVar
from typing_extensions import ParamSpec

P = ParamSpec('P')
T = TypeVar('T')

 # We reuse 'P' in the return type, but replace 'T' with 'str'
def stringify(func: Callable[P, T]) -> Callable[P, str]:
    def wrapper(*args: P.args, **kwds: P.kwargs) -> str:
        return str(func(*args, **kwds))
    return wrapper

 @stringify
 def add_forty_two(value: int) -> int:
     return value + 42
```

Comment (JasonGrace2282) [reply manim/mobject/opengl/opengl_mobject.py:109]: Interesting, didn't realize `ParamSpec` had been backported to `typing_extensions`!

**Comment by JasonGrace2282** [comment]
Is this still part of the initial plan for experimental? If not, I would prefer to close this and add decorators again after experimental is ready instead of dealing with merge conflicts.

**Pull request closed without merging**